### PR TITLE
Introduces unified validation result structure

### LINKED ISSUE #45: Consider "actual" vs "real" terminology
Per @kylef's suggestion:

> Most similiar assertion (test) style functions both in NodeJS (https://nodejs.org/api/all.html#assert_assert_equal_actual_expected_message), chai (https://www.chaijs.com/api/bdd/#method_fail), etc use the terminology "expected" and "actual".
>
> Would it make sense to rename real to actual for consistency?

### LINKED ISSUE #47: Consider requirement keywords from RFC2119
Per @kylef's suggestion:

> As soon as I read this with capital NOT I couldn't stop thinking about it missing SHOULD/MUST/SHALL from [RFC2119](https://tools.ietf.org/html/rfc2119), perhaps something to consider using:
>
> https://github.com/apiaryio/gavel-spec/pull/44#discussion_r293300165

### LINKED ISSUE #48: Define unified Gavel validation result shape
We must define a unified validation result Object structure for Gavel.

## Motivation

1. To become independent from the JSON validators we use or may use. Any validator's output is coerced to the unified validation result structure, preventing us from breaking changes.
1. To contain common data that is currently missing (i.e. `actual`/`expected` values of individual fields, normalize pointers format, adjust fields results to be easily printed out).

### PR DESCRIPTION
**Pull request by artem-zakharchenko**
BREAKING CHANGE: Gavel validation result structure includes the next changes.

## GitHub

- Closes #48 
- Closes #47 
- Closes #45

### Features

- Validity of an HTTP message and HTTP message fields is stored in the `valid` property
```diff
- result.isValid
+result.valid
-result.fields.body.isValid
+result.fields.body.valid
```

- Each HTTP message field contains the `kind` enum property describing the end value type:
```js
result.fields.body.kind // "json"
result.fields.statusCode.kind // "text"
```

- Each HTTP message field error now contains a `location` property that stores information about the occurred error, specific to the `kind` value:
```ts
interface Error {
  message: string
  location?: {
    pointer: string
    property: string[]
  }
}
```

### Deprecations

- Each HTTP message field no longer contains the next properties:
  - `validator`
  - `expectedType`
  - `realType`
  - `rawData`

### CODE REVIEW

**Comment by artem-zakharchenko** [comment]
I would also take this opportunity to refine certain specification aspects:

* Use common requirements vocabulary (#47)
* Refine the way spec features are declared, make sure we test behavior, not implementation

File: features/expectations/body_text_example.feature [hunk 1]
```diff
@@ -46,6 +46,5 @@ Feature: Body - text example
       Orange, strawberry, banana?
       Dog, cat, mouse!
       """
-    Then field "body" is valid
-    And Request or Response is valid
-
+    Then field "body" MUST valid
```
Comment (honzajavorek) [inline features/expectations/body_text_example.feature]:
```suggestion
    Then field "body" MUST be valid
```

File: features/expectations/body_text_example.feature [hunk 2]
```diff
@@ -46,6 +46,5 @@ Feature: Body - text example
...
+    And Request or Response MUST valid
```
Comment (honzajavorek) [inline features/expectations/body_text_example.feature]:
```suggestion
    And Request or Response MUST be valid
```

File: features/javascript/response.feature [hunk 3]
```diff
@@ -6,33 +6,33 @@ Feature: Response validation
...
+      "body": "{\n  \"origin\": \"94.113.241.2\"\n}",
     };
     """
-    And you define following expected HTTP response object:
+    And you define following HTTP response object:
     """
-    const expected = {
+    const real = {
       "statusCode": "200",
+      "statusMessage": "OK",
```
Comment (honzajavorek) [inline features/javascript/response.feature]: Do we want to have it in the examples if Gavel doesn't validate it?

File: features/validators.feature [hunk 4]
```diff
@@ -5,11 +5,7 @@ Feature: Data validators and output format
   Scenario: Validators output format
```
Comment (honzajavorek) [inline features/validators.feature]: I assume this file is still WIP?

File: features/javascript/response.feature (hunk 3 above)
Comment (artem-zakharchenko) [reply features/javascript/response.feature]: This must be removed.

**Comment by artem-zakharchenko** [comment]
> as essentially everything got transformed to MUST or MUST NOT

This is because currently we don't state optional behaviors. For example, when describing a `result.fields[name].errors[n].location` property we can use "MAY" as it's exactly an optional property. We cannot write a strict assertion because it would fail in cases where `location` is not set and it's perfectly valid that way. It brings me to the question of _how_ we write a spec. At the moment I wouldn't say it's evolving around features, as neither "request" not "response" are features of Gavel. Perhaps, we should reconsider how we describe the features.

I imagine the features are:

* `validate()` - high-level definition of how validation behaves. Must include failing scenarios too (currently only positives).
* validating `[fieldName]` - detailed explanation how validation of individual fields is performed. Describes required and optional properties, and explains when those are expected.

I would like to think about feature descriptions like docs (which they are, in fact). So the docs must mention the returned structures, and explain when and why certain properties are present or not. For example, currently the `errors[n]` properties are described like a feature description:

https://github.com/apiaryio/gavel-spec/blob/0cb8b323a7f5832852c5be32d6feab42f177a589/features/validators.feature#L8-L12

I suggest we integrate them into actual feature branches to have more concrete examples when certain properties appear and what they contain.

> I suggest we use "HTTP message" instead of "Request or Response" where both are possible

Absolutely and extremely agree.

> Perhaps "you have following expected HTTP request" would make more sense?

I don't see a reason why not. It does move away from the emphasis on _define_, which is not often the case (as you've mentioned).

File: features/validators.feature (hunk 4 above)
Comment (artem-zakharchenko) [reply features/validators.feature]: Yes. I wanted to discuss whether we need this feature, as it's not a Gavel's feature, but rather test suites for Gavel internals. These belong more to unit tests of Gavel than to gavel-spec. 

Instead, we should write feature files for separate fields (`body.feature`, `statusCode.feature`) to express how the validation of those fields must go.

What do you think?

Comment (artem-zakharchenko) [reply features/validators.feature]: Also I think this block is too technical to be in a feature description:

```gherkin
Given you call:
    """
    const gavel = require('gavel');
    """
```

It also couples the feature descriptions with the way the library is exported.

**Comment by artem-zakharchenko** [comment]
@honzajavorek, also shouldn't Gavel CLI start to return the validation result object instead of terminating/not terminating the process? Process termination alone is not a reliable truth to assume that given HTTP messages are invalid/valid, as it may happen for various reasons. Especially when we started to introduce exceptions in Gavel (i.e. on malformed JSON Schema), the nature of process termination reasons expanded.

I advocate Gavel CLI returns the validation result Object for two reasons:

1. To align CLI and NodeJS usage expectations
1. Because validation result is already known

File: features/validators.feature (hunk 4 above)
Comment (honzajavorek) [reply features/validators.feature]: I agree. The "feature" here shouldn't be validators, but e.g. "body validation".

**Comment by honzajavorek** [comment]
> shouldn't Gavel CLI start to return the validation result object instead of terminating/not terminating the process

Yeah, that's a valid point. My concern is I don't know how many people actually use that CLI. I think it's too cute to be killed, but I'm not convinced it is cute enough to be worked on/extended. Can we have this as a GitHub issue for future generations and wait if it gets any traction?

> This is because currently we don't state optional behaviors. For example, when describing a `result.fields[name].errors[n].location` property we can use "MAY" as it's exactly an optional property. We cannot write a strict assertion because it would fail in cases where `location` is not set and it's perfectly valid that way. It brings me to the question of _how_ we write a spec. At the moment I wouldn't say it's evolving around features, as neither "request" not "response" are features of Gavel. Perhaps, we should reconsider how we describe the features.

If `location` is optional property and we're about to assert it, then the feature should have two scenarios, one which illustrates on examples under which circumstances the property is there, and another one, which illustrates on examples under which circumstances the property is missing.

> I would like to think about feature descriptions like docs...
>
> So the docs must mention the returned structures, and explain when and why certain properties are present or not...

**TL;DR** I think we should have features per field. I'm not so sure we need to be specific about the properties in Gherkins (in Gavel.js [reference](https://www.divio.com/blog/documentation/) docs, sure). It might make sense to be specific if we want to make sure any implementation of the spec produces the same result structure, but since it's just about Gavel.js now, we're good covering specific properties in the unit/integration tests and be more vague here. But I don't have very strong opinions on this, hence the "mmmmmh-there's-no-rule-of-thumb-I-don't-know-wall-of-text" below (sorry for that).

<details>

--------

I understand features as docs (and tests) of how the thing is expected to behave. That means, I think "body validation" is a feature of Gavel, and the Gherkin file could describe in scenarios (illustrate on examples) how the validation is done - positive scenario, negative scenario, and explain corner cases of the behavior.

I don't think it necessarily needs to describe the input or output in details, unless it's useful (the explicitness of the input/output examples makes it more readable) or unless it describes public interface users directly interact with. But still, even in that case I'm more keen to describe things in abstract way, closer to the behavior itself than to implementation - e.g. buttons and form controls are a public interface user directly interacts with, but I prefer "when I send an email" rather than "when I click the Send button". I think all properties etc. definitely need to be in [reference](https://www.divio.com/blog/documentation/) docs, but don't necessarily need to be directly a visible part of the features.

But in the end it really depends on what the "feature" in hand is. If the "feature" is more abstract, like sending an email, the scenario steps probably shouldn't be very specific regarding what is being clicked on or which property is where. If the "feature" is closer to implementation, like testing the button itself, then it's okay to have a scenario about whether it's red or green and how it changes its text.

In the particular case of Gavel validation result properties I think we should have features at least per one field (i.e. body validation, status code validation, etc.) with scenarios describing the possible situations (especially the corner ones, like validating text vs. json). Whether to directly mention the properties on output, I don't know. I'm fine having the steps saying just "isn't valid" or "it gets validated as (text|json)" where the underlying code inspects the `kind` property.

Thought exercise: If Gavel had more implementations then just Gavel.js, it might make sense to get specific about the public interface, but since the spec doesn't give you steps pre-implemented, like `dredd-hooks-template` does, it actually doesn't really matter. Implementation authors provide both implementation code and the steps code. As far as they comply with the behavior described, it is fine, regardless what properties they use to represent the result. Depends whether the spec wants to make sure the output is uniform for all implementations.
</details>

File: features/expectations/body_text_example.feature (hunk 1 above)
Comment (artem-zakharchenko) [reply features/expectations/body_text_example.feature]: Assertion vocabulary replaced to use more readable `is valid` and `is NOT valid` terms (as previously).

File: features/validators.feature (hunk 4 above)
Comment (artem-zakharchenko) [reply features/validators.feature]: Separate fields validations features have been added.

**Comment by artem-zakharchenko** [comment]
I've went through all the points and our design session, and the current state should reflect it.

@honzajavorek, could you please review it once more? Thank you.

> Gavel's [pull request](https://github.com/apiaryio/gavel.js/pull/221) is also ready according to this spec.

File: features/expectations/bodyJsonExample.feature [hunk 5]
```diff
@@ -2,7 +2,7 @@
 Feature: Body - JSON example
 
   Background:
-    Given you define expected HTTP body using the following "JSON example":
+    Given you expect field "body" to equal:
```
[the remark this answers is not in the document]
Comment (artem-zakharchenko) [reply features/expectations/bodyJsonExample.feature]: I like the `Given` phrasing, but regarding omitting fields: I've tried to have a lego vocabulary so different assertions kinds do not need to introduce new steps. That way I've managed to delete the entire `step_defs` folder and stick up with just 3 small modules to describe steps without losing value from the tests themselves.

This means that the `result field "name" is NOT valid` and `result field "name" equals:` operate on the same vocabulary. If I omit the `field` or `result` part then assertions of the result will be unclear:

```gherkin
Then the HTTP message is NOT valid
And the "body" equals:
"""
# Assertion here
"""
```

I am specifically confused by `the "body" equals` part. It reads like it relates to the "body" I have declared in the scenario expectations, but it's actually a _result_ body.

As a middle ground we could separate these two steps into:

```
- the "fieldName" is valid # simple field assertion
- the validation result for "fieldName" is: # asserts the result block
```

Do you find this solving the problem of reading in reverse?

File: features/expectations/method.feature [hunk 6]
```diff
@@ -1,15 +1,17 @@
 @javascript @stable
-Feature: Method
+Feature: method
```
Comment (honzajavorek) [inline features/expectations/method.feature]:
```suggestion
Feature: Method
```

File: features/expectations/statusCode.feature [hunk 7]
```diff
@@ -0,0 +1,18 @@
+@javascript @stable
+Feature: statusCode
```
Comment (honzajavorek) [inline features/expectations/statusCode.feature]:
```suggestion
Feature: Status Code
```

File: features/expectations/statusCode.feature [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+
+  Background:
+    Given you expect "statusCode" to be "200"
+
+  Scenario: Response status code match
```
Comment (honzajavorek) [inline features/expectations/statusCode.feature]:
```suggestion
  Scenario: Response status code matches
```

File: features/javascript/fields/body.feature [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+            "pointer": "/lastName",
+            "property": ["lastName"]
+          }
+        }
+      ]
+    }
+    """
+
+  # Comparing JSON to plain text produces a validation error
+  # and returns "null" kind, indicating no validation has happened.
```
Comment (honzajavorek) [inline features/javascript/fields/body.feature]: I think this can be description inside the scenario - https://relishapp.com/cucumber/cucumber/docs/gherkin/using-descriptions-to-give-features-context#everything-with-a-description

```gherkin
Scenario: Other body type
  
  Comparing JSON to plain text produces a validation error...

  Given ...
```

File: features/javascript/fields/body.feature [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+            "expected": "{\n  \"firstName\": \"John\",\n  \"lastName\": \"Locke\"\n}",
+            "actual": "Textual body"
+          }
+        }
+      ]
+    }
+    """
+
+    # Comparing text to JSON produces a validation error
+    # and "kind: null" stating that no validation has happened.
```
Comment (honzajavorek) [inline features/javascript/fields/body.feature]: ditto scenario description

File: features/javascript/fields/bodySchema.feature [hunk 11]
```diff
@@ -0,0 +1,108 @@
+@javascript @stable
+Feature: Body validation using JSON Schema (V4)
```
Comment (honzajavorek) [inline features/javascript/fields/bodySchema.feature]:
```suggestion
Feature: Body validation using JSON Schema (draft v4)
```

File: features/javascript/fields/bodySchemaLegacy.feature [hunk 12]
```diff
@@ -0,0 +1,123 @@
+@javascript @stable
+Feature: Body validation using JSON Schema (V3)
```
Comment (honzajavorek) [inline features/javascript/fields/bodySchemaLegacy.feature]:
```suggestion
Feature: Body validation using JSON Schema (draft v3)
```

File: features/javascript/fields/method.feature [hunk 13]
```diff
@@ -0,0 +1,47 @@
+@javascript @stable
```
Comment (honzajavorek) [inline features/javascript/fields/method.feature]: I'm wondering whether the `@stable` tag is good for anything... do you happen to know whether we use it anywhere?

File: features/javascript/validate.feature [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+        "accept": "*/*"
+      },
+      "body": ""
+    }
+    """
+    When Gavel validates the HTTP message
+    # When you call:
+    # """
+    # gavel.validate(expectedRequest, actualRequest)
+    # """
```
Comment (honzajavorek) [inline features/javascript/validate.feature]: Feels like in this case it should be the step?

```gherkin
When I call gavel.validate(expected, actual)
```

To keep it DRY, you can easily map both "When you call gavel.validate(expected, actual)" and "When Gavel validates the HTTP message" to the same handling function so they're equivalent, but I wouldn't shy away from writing the scenario with the clearest words possible in the context.

Comment (honzajavorek) [reply features/javascript/validate.feature]: Note: it is a common practice to prefer "I" over "you" in case there's a person involved in the scenario. It helps with the empathy when reading.

File: features/javascript/validate.feature [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+      "date": "Wed, 03 Jul 2013 13:30:53 GMT",
+      "server": "gunicorn/0.17.4",
+      "content-length": "30",
+      "connection": "keep-alive"
+    },
+    "body": "{\n  \"origin\": \"94.113.241.2\"\n}"
+  }
+  """
+  When Gavel validates the HTTP message
+  # When you call:
```
Comment (honzajavorek) [inline features/javascript/validate.feature]: ditto make this the When

File: features/javascript/validate.feature [hunk 16]
```diff
@@ -0,0 +1,159 @@
+@stable @javascript
+Feature: validate
+
+  Scenario: HTTP Request
+    Given you expect the following HTTP Request:
```
Comment (honzajavorek) [inline features/javascript/validate.feature]: ditto my note below, `I` is preferred over `you` (may apply to other places I've could overlook).

File: features/javascript/fields/method.feature (hunk 13 above)
Comment (artem-zakharchenko) [reply features/javascript/fields/method.feature]: Gavel.js only runs those tests which are marked with the `@stable` tag. I believe it was meant to allow `@proposal` and `@draft` in the spec, without affecting the CI of the implementation (i.e. `gavel`).

Comment (honzajavorek) [reply features/javascript/fields/method.feature]: Ah. I don't think we need this kind of granularity now. Having proposals in Pull Requests seem okay for our needs. No priority in dropping the label though, up to you.

Comment (artem-zakharchenko) [reply features/javascript/fields/method.feature]: `@stable` tag removed in 411239f34a86fe364ec202db9a60ebaf9381a3e9.

I've removed any mention of it, including the "add-implementation" guidelines. Those guidelines, however, look quite obsolete and I would propose to change/remove them as they do not depict the actual process of contributing to the spec anymore (no language-specific implementations, no tags).

File: features/javascript/validate.feature (hunk 14 above)
Comment (artem-zakharchenko) [reply features/javascript/validate.feature]: Thanks, I've added a new step definition and mapped it to the existing one.

File: features/javascript/validate.feature (hunk 15 above)
Comment (artem-zakharchenko) [reply features/javascript/validate.feature]: Mapped. Thanks.

**Comment by artem-zakharchenko** [comment]
Thanks for the review! All the comments addressed.

File: features/javascript/fields/method.feature (hunk 13 above)
Comment (honzajavorek) [reply features/javascript/fields/method.feature]: Agreed
